dhis2/dhis2-core
# feat: improved data integrity execution [DHIS2-16223]

### Summary
Various improvements around the data integrity execution.

* details jobs run as a new job type `DATA_INTEGRITY_DETAILS` so they are not blocked by a summary still running  
* the execution time of jobs is remembered to allow sorting them fast to slow when executing a set of checks
* `slow` checks always run last (unless we know their execution time and it is fast)
* checks maintain a transient `averageExecutionTime` property that is also readable in the API
* checks have a `isProgrammatic` flag if they are not SQL based
* all programmatic checks are considered `slow` now (because of that `*` now also expands to slow tests)
* SQL based checks take precedence over programmatic checks of the same name 

### API
The GET `/api/dataIntegrity` endpoint has two added filters
* `slow`: e.g. `/api/dataIntegrity?slow=true` to get only slow checks
* `programmatic`: e.g. `/api/dataIntegrity?programmatic=false` to get only SQL based checks

The new filters can be combined with each other and with the existing `section` filter (AND logic).

### Automatic Testing
New unit tests were added to tests the execution time tracking in `DataIntegrityCheck` as well as the fast to slow sorting.

New integration tests were added to test the added `slow` and `programmatic` filters for the API.

The execution of the new `DATA_INTEGRITY_DETAILS` job type is already covered by many tests that do details checks.

### Manual Testing
Testing filters, try different combinations and observe the result items, e.g.

* `/api/dataIntegrity?slow=true` vs `/api/dataIntegrity?slow=true` vs `/api/dataIntegrity`
* `/api/dataIntegrity?programmatic=true` vs `/api/dataIntegrity?programmatic=true` vs `/api/dataIntegrity`
* `/api/dataIntegrity?programmatic=true&section=Categories`
